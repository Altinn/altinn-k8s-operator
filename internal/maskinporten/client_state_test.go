package maskinporten

import (
	"context"
	"encoding/json"
	"testing"
	"time"

	"github.com/altinn/altinn-k8s-operator/internal/crypto"
	"github.com/altinn/altinn-k8s-operator/internal/operatorcontext"
	"github.com/altinn/altinn-k8s-operator/test/utils"
	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/go-jose/go-jose/v4"
	"github.com/jonboulle/clockwork"
	. "github.com/onsi/gomega"
)

func TestPublicJwksConversion(t *testing.T) {
	g := NewWithT(t)

	jwks, err := createTestJwks()
	g.Expect(err).NotTo(HaveOccurred())
	g.Expect(jwks).NotTo(BeNil())
	g.Expect(jwks.Keys[0].Certificates).NotTo(BeNil())

	publicJwks, err := getPublicOnlyJwks(jwks)
	g.Expect(err).NotTo(HaveOccurred())
	g.Expect(publicJwks).NotTo(BeNil())
	// Certificates is marshalled as "x5c", which Maskinporten doens't want
	g.Expect(publicJwks.Keys[0].Certificates).To(BeNil())
	g.Expect(jwks.Keys[0].Certificates).NotTo(BeNil())

	json, err := json.Marshal(publicJwks)
	g.Expect(err).NotTo(HaveOccurred())
	g.Expect(json).NotTo(BeNil())
	snaps.MatchJSON(t, json)

	// TODO: assert that private key fields are not present in JWK
}

func createTestJwks() (*jose.JSONWebKeySet, error) {
	operatorContext := operatorcontext.DiscoverOrDie(context.Background())
	clock := clockwork.NewFakeClockAt(time.Date(2024, 1, 1, 0, 0, 0, 0, time.UTC))
	random := utils.NewDeterministicRand()
	service := crypto.NewService(operatorContext, clock, random)

	appId := "app1"
	jwks, err := service.CreateJwks(appId)
	if err != nil {
		return nil, err
	}

	return jwks, nil
}
