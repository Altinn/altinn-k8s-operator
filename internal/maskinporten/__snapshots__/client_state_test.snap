
[TestPublicJwksConversion - 1]
{
 "keys": [
  {
   "alg": "RS256",
   "e": "AQAB",
   "kid": "b67efbf1-f75d-4fb1-9cac-242cc2b5d334.0",
   "kty": "RSA",
   "n": "tabD3r_8nii4gPerNyupCXWm9baeeylz6ZoAeBREKTY2hRimFZgWvDsuMzReGqijrsijmkBxkScQpCW5fGvhGCgw0KUE5ucIYlmV_CIA_ZwHtpwc_nuqXSqtWBY59wSE8TH3Vm9pk-rljBGpxnb8qBXQ50o_CsPDwDuEK_lUll_WFOAx5nL5qcffiabEiGnl2oNnQaXhtbPqCIDufSdpfaPHnlz_ko6klbs1fewxpM4zFxvDZLVnxDSqYgi1DZVNcWH7lPvHvo7y7rbs698VXs-uztTJ4fZ5pm-epqwKnagSeV-9f_vchS09qp3A9G7FeQWucZLHqfEver2x6C2nNw",
   "use": "sig"
  }
 ]
}
---
